tests/unit_tests/percolator/UnitTest_Percolator_SetHandler.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2021 Brian Raiter <breadbox@muppetlabs.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <gtest/gtest.h>
+#include <sstream>
+#include "SetHandler.h"
+
+/* A simple class that tracks global deletions.
+ */
+class DeletionTracker {
+  public:
+    DeletionTracker() { }
+    virtual ~DeletionTracker() { ++deletionCount; }
+    static void reset() { deletionCount = 0; }
+    static int deletionCount;
+};
+int DeletionTracker::deletionCount = 0;
+
+/* A SetHandler subclass with access to some protected methods.
+ */
+class SetHandlerEx : public SetHandler {
+  public:
+    SetHandlerEx(int maxPSMs) : SetHandler(maxPSMs) { }
+    virtual ~SetHandlerEx() { }
+    int getOptionalFieldsEx(const std::string& headerLine, 
+                            std::vector<OptionalField>& optionalFields) {
+        return this->getOptionalFields(headerLine, optionalFields);
+    }
+};
+
+/*
+ * Unit tests for the SetHandler classes.
+ */
+
+class SetHandlerTest : public ::testing::Test {
+  protected:
+    int testInput(SetHandler *pSH, char const *input);
+};
+
+/* For tracking DataSet destructions.
+ */
+class TrackedDataSet : public DataSet, DeletionTracker { };
+
+// Given a string containing input data, wrap it in a stream object
+// and feed it to SetHandler::readTab().
+int SetHandlerTest::testInput(SetHandler *handler, char const *input)
+{
+    std::istringstream str(input);
+    SanityCheck *pCheck = NULL;
+    int n = handler->readTab(str, pCheck);
+    if (n > 0)
+        EXPECT_TRUE(pCheck != NULL);
+    delete pCheck;
+    return n;
+}
+
+// Verify DataSets are stored correctly, and are deleted when
+// requested.
+TEST_F(SetHandlerTest, TestSubsetStoring)
+{
+    SetHandler sh(0);
+    TrackedDataSet *pos = new TrackedDataSet();
+    TrackedDataSet *neg = new TrackedDataSet();
+    pos->setLabel(+1);
+    neg->setLabel(-1);
+    DeletionTracker::reset();
+    sh.push_back_dataset(pos);
+    sh.push_back_dataset(neg);
+    EXPECT_EQ(0, DeletionTracker::deletionCount);
+    EXPECT_EQ(pos, sh.getSubsetFromLabel(+1));
+    EXPECT_EQ(neg, sh.getSubsetFromLabel(-1));
+
+    sh.reset();
+    EXPECT_EQ(2, DeletionTracker::deletionCount);
+}
+
+// Validate basic parsing of optional fields.
```

Comment:
Differently how? I'm not sure I'm aware of this.

Reply:
Sorry, I replied in the wrong place... This was my reply:

I think it stems from this:

https://github.com/percolator/percolator/wiki/Retention-time-and-calibration

So now, for instance, I have retention time in the third column, and now it
parsed as an optional header. But I think fi it would had another position
it would be read as an feature header.